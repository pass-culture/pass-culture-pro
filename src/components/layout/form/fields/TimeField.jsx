/* eslint
  react/jsx-one-expression-per-line: 0 */
import classnames from 'classnames'
import React from 'react'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'
import { composeValidators } from 'react-final-form-utils'
import ReactTimeInput from 'react-time-input'

import validateRequiredField from 'components/layout/form/utils/validateRequiredField'

export const TimeField = ({
  autoComplete,
  className,
  clearable,
  disabled,
  id,
  label,
  locale,
  name,
  placeholder,
  readOnly,
  renderValue,
  required,
  type,
  validate,
  tz,
  // https://www.npmjs.com/package/react-time-input
  ...ReactTimeInputProps
}) => {
  const requiredValidate =
    required && typeof required === 'function'
      ? required
      : (required && validateRequiredField) || undefined

  return (
    <Field
      name={name}
      validate={composeValidators(validate, requiredValidate)}
      render={({ input, meta }) => {
        return (
          <div
            className={classnames('field time-field', className, {
              'is-read-only': readOnly,
            })}
            id={id}>
            <label
              htmlFor={name}
              className={classnames('field-label', { empty: !label })}>
              {label && (
                <span>
                  <span>{label}</span>
                  {required && !readOnly && (
                    <span className="field-asterisk">*</span>
                  )}
                </span>
              )}
            </label>
            <div className="field-control">
              <div className="field-value flex-columns items-center">
                <div className="field-inner flex-columns items-center">
                  {readOnly ? (
                    <input
                      className="field-input field-time"
                      readOnly
                      value={input.value}
                    />
                  ) : (
                    <ReactTimeInput
                      className="field-input field-time"
                      initTime={input.value}
                      {...input}
                      {...ReactTimeInputProps}
                      onTimeChange={time => input.onChange(time)}
                    />
                  )}
                </div>
                {renderValue()}
              </div>
            </div>
          </div>
        )
      }}
    />
  )
}

TimeField.defaultProps = {
  autoComplete: false,
  className: '',
  disabled: false,
  id: null,
  label: '',
  locale: 'fr',
  placeholder: 'Please enter a value',
  readOnly: false,
  renderValue: () => null,
  required: false,
  validate: null,
}

TimeField.propTypes = {
  autoComplete: PropTypes.bool,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  label: PropTypes.string,
  locale: PropTypes.string,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  renderValue: PropTypes.func,
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
  validate: PropTypes.func,
}

export default TimeField
