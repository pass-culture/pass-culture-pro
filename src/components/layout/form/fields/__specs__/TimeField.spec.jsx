/* eslint-disable react/jsx-no-bind */
import { mount } from 'enzyme'
import React from 'react'
import { Form } from 'react-final-form'

import TimeField from '../TimeField'

describe('src | components | layout | form | TimeField', () => {
  it('should submit a form with a time', async () => {
    await new Promise(done => {
      // given
      const wrapper = mount(
        <Form
          onSubmit={handleOnSubmit}
          render={({ handleSubmit }) => (
            <form>
              <TimeField name="myTime" />
              <button
                onClick={handleSubmit}
                type="submit"
              >
                {'Submit'}
              </button>
            </form>
          )}
        />
      )

      // when
      wrapper
        .find(TimeField)
        .find('input[name="myTime"]')
        .simulate('change', { target: { value: '03:45' } })
      wrapper.find('button[type="submit"]').simulate('click')

      // then
      function handleOnSubmit(formValues) {
        expect(formValues.myTime).toStrictEqual('03:45')
        done()
      }
    })
  })
})
