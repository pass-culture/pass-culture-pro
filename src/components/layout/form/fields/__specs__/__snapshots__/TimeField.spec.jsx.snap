// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | layout | form | TimeField should match the snapchot 1`] = `ShallowWrapper {}`;
